src/dmd/typesem.d

```diff
@@ -4151,13 +4151,75 @@ Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
                 /* Rewrite as:
                  *  this.d
                  */
-                if (hasThis(sc))
+                AggregateDeclaration ad = d.isMember4();
+                if (auto f = hasThis(sc))
                 {
                     // This is almost same as getRightThis() in expression.c
-                    Expression e1 = new ThisExp(e.loc);
+                    Expression e1;
+                    Type t;
+                    /* returns: true to continue, false to return */
+                    bool skipNestedFunctions(Dsymbol s)
+                    {
+                        while (s && s.isFuncDeclaration())
+                        {
+                            FuncDeclaration f = s.isFuncDeclaration();
+                            if (f.vthis)
+                            {
+                                if (f.vthis2)
+                                {
```

Comment:
this ast manipulation looks familiar maybe it can be factored into a function?

Reply:
done